File: src/main/java/thedarkcolour/futuremc/entity/drowned/EntityDrowned.java
package thedarkcolour.futuremc.entity.drowned;

import com.google.common.base.Predicate;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.IRangedAttackMob;
import net.minecraft.entity.SharedMonsterAttributes;
import net.minecraft.entity.ai.EntityAIAttackRanged;
import net.minecraft.entity.ai.EntityAIBase;
import net.minecraft.entity.ai.EntityAIHurtByTarget;
import net.minecraft.entity.ai.EntityAIMoveToBlock;
import net.minecraft.entity.ai.EntityAINearestAttackableTarget;
import net.minecraft.entity.ai.EntityAIWander;
import net.minecraft.entity.ai.EntityAIZombieAttack;
import net.minecraft.entity.ai.EntityMoveHelper;
import net.minecraft.entity.ai.RandomPositionGenerator;
import net.minecraft.entity.monster.EntityIronGolem;
import net.minecraft.entity.monster.EntityZombie;
import net.minecraft.entity.passive.EntityVillager;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.pathfinding.Path;
import net.minecraft.pathfinding.PathNavigateGround;
import net.minecraft.pathfinding.PathNavigateSwimmer;
import net.minecraft.pathfinding.PathNodeType;
import net.minecraft.pathfinding.PathPoint;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;
import thedarkcolour.futuremc.entity.trident.EntityTrident;
import thedarkcolour.futuremc.init.Init;
import thedarkcolour.futuremc.sound.Sounds;

import java.util.Random;

public class EntityDrowned extends EntityZombie implements IRangedAttackMob {
    private boolean field_204718_bx;
    private PathNavigateSwimmer waterNavigator;
    private PathNavigateGround groundNavigator;

    public EntityDrowned(World worldIn) {
        super(worldIn);
        this.stepHeight = 1.0F;
        this.moveHelper = new EntityDrowned.MoveHelper(this);
        this.setPathPriority(PathNodeType.WATER, 0.0F);
        this.waterNavigator = new PathNavigateSwimmer(this, worldIn);
        this.groundNavigator = new PathNavigateGround(this, worldIn);
    }

    @Override
    protected void applyEntityAI() {
        this.tasks.addTask(1, new EntityDrowned.AIGoToWater(this, 1.0D));
        this.tasks.addTask(2, new EntityDrowned.AITridentAttack(this, 1.0D, 40, 10.0F));
        this.tasks.addTask(2, new EntityDrowned.AIAttack(this, 1.0D, false));
        this.tasks.addTask(5, new EntityDrowned.AIGoToBeach(this, 1.0D));
        this.tasks.addTask(6, new EntityDrowned.AISwimUp(this, 1.0D, this.world.getSeaLevel()));
        this.tasks.addTask(7, new EntityAIWander(this, 1.0D));
        this.targetTasks.addTask(1, new EntityAIHurtByTarget(this, true, EntityDrowned.class));
        this.targetTasks.addTask(2, new EntityAINearestAttackableTarget<>(this, EntityPlayer.class, 10, true, false, new AttackTargetPredicate(this)));
        this.targetTasks.addTask(3, new EntityAINearestAttackableTarget<>(this, EntityVillager.class, false));
        this.targetTasks.addTask(3, new EntityAINearestAttackableTarget<>(this, EntityIronGolem.class, true));
    }

    public boolean shouldAttack(EntityLivingBase p_204714_1_) {
        if (p_204714_1_ != null) {
            return !this.world.isDaytime() || p_204714_1_.isInWater();
        } else {
            return false;
        }
    }

    private boolean func_204715_dF() {
        if (this.field_204718_bx) {
            return true;
        } else {
            EntityLivingBase entitylivingbase = this.getAttackTarget();
            return entitylivingbase != null && entitylivingbase.isInWater();
        }
    }

    public void func_204713_s(boolean p_204713_1_) {
        this.field_204718_bx = p_204713_1_;
    }

    @Override
    public void attackEntityWithRangedAttack(EntityLivingBase target, float distanceFactor) {
        EntityTrident entitytrident = new EntityTrident(this.world, this, new ItemStack(Init.TRIDENT));

        double d0 = target.posX - this.posX;
        double d1 = target.getEntityBoundingBox().minY + (double)(target.height / 3.0F) - entitytrident.posY;
        double d2 = target.posZ - this.posZ;
        double d3 = MathHelper.sqrt(d0 * d0 + d2 * d2);
        entitytrident.shoot(d0, d1 + d3 * (double)0.2F, d2, 1.6F, (float)(14 - this.world.getDifficulty().getId() * 4));
        this.playSound(Sounds.TRIDENT_THROW, 1.0F, 1.0F / (this.getRNG().nextFloat() * 0.4F + 0.8F));
        this.world.spawnEntity(entitytrident);
    }

    @Override
    public void setSwingingArms(boolean swingingArms) {

    }

    protected boolean isCloseToPathTarget() {
        Path path = this.getNavigator().getPath();
        if (path != null) {
            PathPoint pathpoint = path.getTarget();
            double d0 = this.getDistanceSq(pathpoint.x, pathpoint.y, pathpoint.z);
            return d0 < 4.0D;
        }
        return false;
    }

    public static class AIGoToWater extends EntityAIBase {
        private final EntityDrowned entityIn;
        private double x, y, z;
        private final double speed;
        private final World world;

        public AIGoToWater(EntityDrowned entityIn, double speed) {
            this.entityIn = entityIn;
            this.speed = speed;
            this.world = entityIn.world;
            this.setMutexBits(1);
        }

        @Override
        public boolean shouldExecute() {
            if (!this.world.isDaytime()) {
                return false;
            }
            else if (this.entityIn.isInWater()) {
                return false;
            }
            else {
                Vec3d vec3d = this.getWaterPos();
                if (vec3d == null) {
                    return false;
                }
                else {
                    this.x = vec3d.x;
                    this.y = vec3d.y;
                    this.z = vec3d.z;
                    return true;
                }
            }
        }

        @Override
        public boolean shouldContinueExecuting() {
            return !this.entityIn.getNavigator().noPath();
        }

        /**
         * Execute a one shot task or start executing a continuous task
         */
        @Override
        public void startExecuting() {
            this.entityIn.getNavigator().tryMoveToXYZ(this.x, this.y, this.z, this.speed);
        }

        private Vec3d getWaterPos() {
            Random random = this.entityIn.getRNG();
            //noinspection ConstantConditions
            BlockPos pos = new BlockPos(this.entityIn.posX, this.entityIn.getCollisionBoundingBox().minY, this.entityIn.posZ);

            for(int i = 0; i < 10; ++i) {
                BlockPos pos1 = pos.add(random.nextInt(20) - 10, 2 - random.nextInt(8), random.nextInt(20) - 10);
                if (this.world.getBlockState(pos1).getBlock() == Blocks.WATER) {
                    return new Vec3d(pos1.getX(), pos1.getY(), pos1.getZ());
                }
            }

            return null;
        }
    }

    public static class AISwimUp extends EntityAIBase {
        private final EntityDrowned drowned;
        private final double speed;
        private final int targetY;
        private boolean obstructed;

        public AISwimUp(EntityDrowned drowned, double p_i48908_2_, int p_i48908_4_) {
            this.drowned = drowned;
            this.speed = p_i48908_2_;
            this.targetY = p_i48908_4_;
        }

        public boolean shouldExecute() {
            return !this.drowned.world.isDaytime() && this.drowned.isInWater() && this.drowned.posY < (double)(this.targetY - 2);
        }

        public boolean shouldContinueExecuting() {
            return this.shouldExecute() && !this.obstructed;
        }

        public void updateTask() {
            if (this.drowned.posY < (double)(this.targetY - 1) && (this.drowned.getNavigator().noPath() || this.drowned.isCloseToPathTarget())) {
                Vec3d vec3d = RandomPositionGenerator.findRandomTargetBlockTowards(this.drowned, 4, 8, new Vec3d(this.drowned.posX, this.targetY - 1, this.drowned.posZ));
                if (vec3d == null) {
                    this.obstructed = true;
                    return;
                }

                this.drowned.getNavigator().tryMoveToXYZ(vec3d.x, vec3d.y, vec3d.z, this.speed);
            }

        }

        public void startExecuting() {
            this.drowned.func_204713_s(true);
            this.obstructed = false;
        }

        public void resetTask() {
            this.drowned.func_204713_s(false);
        }
    }

    public static class AITridentAttack extends EntityAIAttackRanged {
        private final EntityDrowned attacker;

        public AITridentAttack(EntityDrowned attacker, double speed, int maxAttackTime, float maxAttackDistanceIn) {
            super(attacker, speed, maxAttackTime, maxAttackDistanceIn);
            this.attacker = attacker;
        }

        public boolean shouldExecute() {
            return super.shouldExecute() && this.attacker.getHeldItemMainhand().getItem() == Init.TRIDENT;
        }

        public void startExecuting() {
            super.startExecuting();
            this.attacker.setSwingingArms(true);
        }

        public void resetTask() {
            super.resetTask();
            this.attacker.setSwingingArms(false);
        }
    }

    public static class AIAttack extends EntityAIZombieAttack {
        private final EntityDrowned drowned;

        public AIAttack(EntityDrowned drowned, double speed, boolean longMemory) {
            super(drowned, speed, longMemory);
            this.drowned = drowned;
        }

        public boolean shouldExecute() {
            return super.shouldExecute() && this.drowned.shouldAttack(this.drowned.getAttackTarget());
        }

        public boolean shouldContinueExecuting() {
            return super.shouldContinueExecuting() && this.drowned.shouldAttack(this.drowned.getAttackTarget());
        }
    }

    static class AIGoToBeach extends EntityAIMoveToBlock {
        private final EntityDrowned drowned;

        AIGoToBeach(EntityDrowned drowned, double p_i48911_2_) {
            super(drowned, p_i48911_2_, 8);
            this.drowned = drowned;
        }

        public boolean shouldExecute() {
            return super.shouldExecute() && !this.drowned.world.isDaytime() && this.drowned.isInWater() && this.drowned.posY >= (double)(this.drowned.world.getSeaLevel() - 3);
        }

        public boolean shouldContinueExecuting() {
            return super.shouldContinueExecuting();
        }

        public void startExecuting() {
            this.drowned.func_204713_s(false);
            this.drowned.navigator = this.drowned.groundNavigator;
            super.startExecuting();
        }

        public void resetTask() {
            super.resetTask();
        }

        protected boolean shouldMoveTo(World worldIn, BlockPos pos) {
            BlockPos blockpos = pos.up();
            return (worldIn.isAirBlock(blockpos) && worldIn.isAirBlock(blockpos.up())) && worldIn.getBlockState(pos).isTopSolid();
        }
    }

    public static class AttackTargetPredicate implements Predicate<EntityPlayer> {
        private final EntityDrowned entity;

        public AttackTargetPredicate(EntityDrowned drowned) {
            this.entity = drowned;
        }

        @Override
        public boolean apply(EntityPlayer playerIn) {
            return this.entity.shouldAttack(playerIn);
        }
    }

    static class MoveHelper extends EntityMoveHelper {
        private final EntityDrowned drowned;

        MoveHelper(EntityDrowned entity) {
            super(entity);
            this.drowned = entity;
        }

        public void onUpdateMoveHelper() {
            EntityLivingBase entitylivingbase = this.drowned.getAttackTarget();
            if (this.drowned.func_204715_dF() && this.drowned.isInWater()) {
                if (entitylivingbase != null && entitylivingbase.posY > this.drowned.posY || this.drowned.field_204718_bx) {
                    this.drowned.motionY += 0.002D;
                }

                if (this.action != EntityMoveHelper.Action.MOVE_TO || this.drowned.getNavigator().noPath()) {
                    this.drowned.setAIMoveSpeed(0.0F);
                    return;
                }

                double d0 = this.posX - this.drowned.posX;
                double d1 = this.posY - this.drowned.posY;
                double d2 = this.posZ - this.drowned.posZ;
                double d3 = MathHelper.sqrt(d0 * d0 + d1 * d1 + d2 * d2);
                d1 = d1 / d3;
                float f = (float)(MathHelper.atan2(d2, d0) * (double)(180F / (float)Math.PI)) - 90.0F;
                this.drowned.rotationYaw = this.limitAngle(this.drowned.rotationYaw, f, 90.0F);
                this.drowned.renderYawOffset = this.drowned.rotationYaw;
                float f1 = (float)(this.speed * this.drowned.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).getAttributeValue());
                this.drowned.setAIMoveSpeed(this.drowned.getAIMoveSpeed() + (f1 - this.drowned.getAIMoveSpeed()) * 0.125F);
                this.drowned.motionY += (double)this.drowned.getAIMoveSpeed() * d1 * 0.1D;
                this.drowned.motionX += (double)this.drowned.getAIMoveSpeed() * d0 * 0.005D;
                this.drowned.motionZ += (double)this.drowned.getAIMoveSpeed() * d2 * 0.005D;
            } else {
                if (!this.drowned.onGround) {
                    this.drowned.motionY -= 0.008D;
                }

                super.onUpdateMoveHelper();
            }
        }
    }
}

File: src/main/java/thedarkcolour/futuremc/entity/drowned/EntityDrowned.kt
package thedarkcolour.futuremc.entity.drowned

import net.minecraft.block.material.Material
import net.minecraft.entity.*
import net.minecraft.entity.ai.*
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.monster.EntityIronGolem
import net.minecraft.entity.monster.EntityZombie
import net.minecraft.entity.passive.EntityVillager
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Biomes
import net.minecraft.init.Items
import net.minecraft.inventory.EntityEquipmentSlot
import net.minecraft.item.ItemStack
import net.minecraft.network.datasync.DataSerializers
import net.minecraft.network.datasync.EntityDataManager
import net.minecraft.pathfinding.PathNavigateGround
import net.minecraft.pathfinding.PathNavigateSwimmer
import net.minecraft.pathfinding.PathNodeType
import net.minecraft.util.DamageSource
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.SoundEvent
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3d
import net.minecraft.world.DifficultyInstance
import net.minecraft.world.EnumDifficulty
import net.minecraft.world.World
import thedarkcolour.core.util.lerp
import thedarkcolour.core.util.stack
import thedarkcolour.futuremc.entity.trident.EntityTrident
import thedarkcolour.futuremc.init.FItems
import thedarkcolour.futuremc.init.Sounds
import kotlin.math.*

class EntityDrowned(world: World) : EntityZombie(world), IRangedAttackMob {
    private val waterNavigator = PathNavigateSwimmer(this, world)
    private val groundNavigator = PathNavigateGround(this, world)
    private var swimmingUp = false
    private var swimAnimation = 0.0f
    private var lastSwimAnimation = 0.0f

    init {
        stepHeight = 1.0f
        moveHelper = MoveHelper(this)
        setPathPriority(PathNodeType.WATER, 0.0f)
    }

    private class MoveHelper(entity: EntityDrowned) : EntityMoveHelper(entity) {
        override fun onUpdateMoveHelper() {
            val target = entity.attackTarget
            entity as EntityDrowned

            if (entity.shouldSwim() && entity.isInWater) {
                if ((target != null && target.posY > entity.posY) || entity.swimmingUp) {
                    entity.posY += 0.002
                }

                if (action != Action.MOVE_TO || entity.navigator.noPath()) {
                    entity.aiMoveSpeed = 0.0f
                }
                val x = posX - entity.posX
                var y = posY - entity.posY
                val z = posZ - entity.posZ
                val distance = sqrt(x * x + y * y + z * z)
                y /= distance
                val f = atan2(x, z) * (180 / PI) - 90.0f
                entity.rotationYaw = limitAngle(entity.rotationYaw, f.toFloat(), 90.0f)
                entity.renderYawOffset = entity.rotationYaw
                val i = speed * entity.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).attributeValue
                val j = lerp(0.125f, entity.aiMoveSpeed, i.toFloat())
                entity.aiMoveSpeed = j
                entity.motionX += j * x * 0.005
                entity.motionY += j * y * 0.1
                entity.motionZ += j * z * 0.005
            } else {
                if (entity.onGround) {
                    entity.motionY -= 0.008
                }
                super.onUpdateMoveHelper()
            }
        }
    }

    override fun getCanSpawnHere(): Boolean {
        val pos = position
        val biome = world.getBiome(pos)
        val flag = world.difficulty != EnumDifficulty.PEACEFUL && isValidLightLevel && world.getBlockState(pos).material == Material.WATER
        return if (biome != Biomes.RIVER && biome != Biomes.FROZEN_RIVER) {
            rand.nextInt(40) == 0 && waterIsDeepEnough(world, pos) && flag
        } else {
            rand.nextInt(15) == 0 && flag
        }
    }

    override fun onInitialSpawn(difficulty: DifficultyInstance, livingdata: IEntityLivingData?): IEntityLivingData? {
        val entityData = super.onInitialSpawn(difficulty, livingdata)
        if (getItemStackFromSlot(EntityEquipmentSlot.OFFHAND).isEmpty && rand.nextFloat() < 0.03) {
            setItemStackToSlot(EntityEquipmentSlot.OFFHAND, FItems.NAUTILUS_SHELL.stack)
            inventoryHandsDropChances[EntityEquipmentSlot.OFFHAND.index] = 2.0f
        }

        return entityData
    }

    private fun waterIsDeepEnough(world: World, pos: BlockPos): Boolean {
        return pos.y < world.seaLevel - 5
    }

    override fun getAmbientSound(): SoundEvent = Sounds.DROWNED_AMBIENT

    override fun getHurtSound(damageSourceIn: DamageSource): SoundEvent = Sounds.DROWNED_HURT

    override fun getDeathSound(): SoundEvent = Sounds.DROWNED_DEATH

    override fun getStepSound(): SoundEvent = Sounds.DROWNED_STEP

    override fun getSwimSound(): SoundEvent = Sounds.DROWNED_SWIM

    override fun getSkullDrop(): ItemStack = ItemStack.EMPTY

    override fun setBreakDoorsAItask(enabled: Boolean) {}

    override fun setEquipmentBasedOnDifficulty(difficulty: DifficultyInstance) {
        if (rand.nextFloat() > 0.9f) {
            if (rand.nextInt(16) < 10) {
                setItemStackToSlot(EntityEquipmentSlot.MAINHAND, FItems.TRIDENT.stack)
            } else {
                setItemStackToSlot(EntityEquipmentSlot.MAINHAND, Items.FISHING_ROD.stack)
            }
        }
    }

    override fun updateEquipmentIfNeeded(itemEntity: EntityItem) {
        val slot = EntityLiving.getSlotForItemStack(itemEntity.item)
        val existing = getItemStackFromSlot(slot)
        val candidate = itemEntity.item
        var flag = false

        if (existing.item == FItems.TRIDENT) {
            if (candidate.item == FItems.TRIDENT) {
                if (candidate.itemDamage < existing.itemDamage) {
                    flag = true
                }
            }
        } else if (candidate.item == FItems.TRIDENT) {
            flag = true
        }

        if (flag) {
            val d0: Double = when (slot.slotType) {
                EntityEquipmentSlot.Type.HAND -> inventoryHandsDropChances[slot.index].toDouble()
                EntityEquipmentSlot.Type.ARMOR -> inventoryArmorDropChances[slot.index].toDouble()
                else -> 0.0
            }
            if (!existing.isEmpty && (rand.nextFloat() - 0.1f).toDouble() < d0) {
                entityDropItem(existing, 0.0f)
            }
            setItemStackToSlot(slot, candidate)
            when (slot.slotType) {
                EntityEquipmentSlot.Type.HAND -> inventoryHandsDropChances[slot.index] = 2.0f
                EntityEquipmentSlot.Type.ARMOR -> inventoryArmorDropChances[slot.index] = 2.0f
                else -> {}
            }
            enablePersistence()
            onItemPickup(itemEntity, candidate.count)
            itemEntity.setDead()
        }
    }

    override fun isNotColliding(): Boolean {
        return world.checkNoEntityCollision(this.entityBoundingBox)
    }

    private var isSwimming
        get() = dataManager[SWIMMING]
        set(value) {
            dataManager[SWIMMING] = value
        }

    override fun isPushedByWater(): Boolean {
        return !isSwimming
    }

    override fun onEntityUpdate() {
        super.onEntityUpdate()

        updateSwimming()
        updateSwimmingAnimation()
    }

    private fun updateSwimming() {
        if (!world.isRemote) {
            if (isServerWorld && isInWater && shouldSwim()) {
                navigator = waterNavigator
                isSwimming = true
            } else {
                navigator = groundNavigator
                isSwimming = false
            }
        }
    }

    private fun updateSwimmingAnimation() {
        lastSwimAnimation = swimAnimation

        swimAnimation = if (swimIsDecreasing()) {
            min(1.0f, swimAnimation + 0.9f)
        } else {
            max(0.0f, swimAnimation - 0.9f)
        }
    }

    private fun swimIsDecreasing(): Boolean {
        return isSwimming || !isElytraFlying
    }

    override fun travel(strafe: Float, vertical: Float, forward: Float) {
        if (isServerWorld && isInWater && shouldSwim()) {
            moveRelative(0.01f, strafe, vertical, forward)
            move(MoverType.SELF, motionX, motionY, motionZ)
            motionX *= 0.9
            motionY *= 0.9
            motionZ *= 0.9
        } else {
            super.travel(strafe, vertical, forward)
        }
    }

    private fun shouldSwim(): Boolean {
        return swimmingUp || attackTarget?.isInWater ?: false
    }

    override fun entityInit() {
        super.entityInit()
        dataManager.register(DROWNING, false)
        dataManager.register(SWINGING_ARMS, false)
        dataManager.register(SWIMMING, false)
    }

    override fun applyEntityAI() {
        tasks.addTask(1, AIGoToWater(this))
        tasks.addTask(2, AITridentAttack(this))
        tasks.addTask(2, AIAttack(this))
        tasks.addTask(5, AIGoToBeach(this))
        tasks.addTask(6, AISwimUp(this))
        tasks.addTask(7, EntityAIWander(this, 1.0))
        targetTasks.addTask(1, EntityAIHurtByTarget(this, true, EntityDrowned::class.java))
        targetTasks.addTask(2, EntityAINearestAttackableTarget(this, EntityPlayer::class.java, 10, true, false, ::shouldAttack))
        targetTasks.addTask(3, EntityAINearestAttackableTarget(this, EntityVillager::class.java, false))
        targetTasks.addTask(3, EntityAINearestAttackableTarget(this, EntityIronGolem::class.java, true  ))
        //targetTasks.addTask(5, EntityAINearestAttackableTarget(this, EntityTurtle::class.java, 10, true, false, ::shouldAttack))
    }

    private class AIGoToWater(val entity: EntityDrowned) : EntityAIBase() {
        private val world = entity.world
        private var x: Double = 0.0
        private var y: Double = 0.0
        private var z: Double = 0.0

        init {
            mutexBits = 0x1
        }

        override fun shouldExecute(): Boolean {
            return if (!world.isDaytime) {
                false
            } else if (entity.isInWater) {
                false
            } else {
                val vec3d = findDestination()
                vec3d?.let { vec ->
                    x = vec.x
                    y = vec.y
                    z = vec.z
                    true
                } ?: false
            }
        }

        private fun findDestination(): Vec3d? {
            val rand = entity.rand
            val pos = BlockPos(entity.posX, entity.entityBoundingBox.minY, entity.posZ)

            for (i in 0..9) {
                val pos1 = pos.add(rand.nextInt(20) - 10, 2 - rand.nextInt(8), rand.nextInt(20) - 10)
                if (world.getBlockState(pos1).material == Material.WATER) {
                    return Vec3d(pos1)
                }
            }

            return null
        }

        override fun startExecuting() {
            entity.navigator.tryMoveToXYZ(x, y, z, 1.0)
        }

        override fun shouldContinueExecuting(): Boolean {
            return !entity.navigator.noPath()
        }
    }

    private class AITridentAttack(val entity: EntityDrowned) : EntityAIAttackRanged(entity, 1.0, 40, 10.0f) {
        override fun shouldExecute(): Boolean {
            return super.shouldExecute() && entity.heldItemMainhand.item == FItems.TRIDENT
        }

        override fun startExecuting() {
            super.startExecuting()
            entity.setSwingingArms(true)
            entity.activeHand = EnumHand.MAIN_HAND
        }

        override fun resetTask() {
            super.resetTask()
            entity.resetActiveHand()
            entity.setSwingingArms(false)
        }
    }

    private class AIAttack(val entity: EntityDrowned) : EntityAIZombieAttack(entity, 1.0, false) {
        override fun shouldExecute(): Boolean {
            return super.shouldExecute() && entity.shouldAttack(entity.attackTarget)
        }

        override fun shouldContinueExecuting(): Boolean {
            return super.shouldContinueExecuting() && entity.shouldAttack(entity.attackTarget)
        }
    }

    private class AIGoToBeach(val entity: EntityDrowned) : EntityAIMoveToBlock(entity, 1.0, 8) {
        override fun shouldMoveTo(worldIn: World, pos: BlockPos): Boolean {
            val up = pos.up()
            return worldIn.isAirBlock(up) && worldIn.isAirBlock(up.up()) && worldIn.getBlockState(pos).isSideSolid(worldIn, pos, EnumFacing.UP)
        }

        override fun shouldExecute(): Boolean {
            return super.shouldExecute() && !entity.world.isDaytime && entity.isInWater && entity.posY >= entity.world.seaLevel - 3
        }

        override fun startExecuting() {
            entity.swimmingUp = false
            entity.navigator = entity.groundNavigator
            super.startExecuting()
        }
    }

    private class AISwimUp(val entity: EntityDrowned) : EntityAIBase() {
        private val targetY = entity.world.seaLevel
        private var obstructed = false

        override fun shouldExecute(): Boolean {
            return !entity.world.isDaytime && entity.isInWater && entity.posY < targetY - 2
        }

        override fun startExecuting() {
            entity.swimmingUp = true
            obstructed = false
        }

        override fun updateTask() {
            if (entity.posY < targetY - 1 && entity.navigator.noPath() || entity.isCloseToPathTarget()) {
                val vec3d = RandomPositionGenerator.findRandomTargetBlockTowards(entity, 4, 8, Vec3d(entity.posX, (targetY - 1).toDouble(), entity.posZ))
                if (vec3d == null) {
                    obstructed = true
                    return
                }

                entity.navigator.tryMoveToXYZ(vec3d.x, vec3d.y, vec3d.z, 1.0)
            }
        }

        override fun shouldContinueExecuting(): Boolean {
            return shouldExecute() && !obstructed
        }

        override fun resetTask() {
            entity.swimmingUp = false
        }
    }

    private fun isCloseToPathTarget(): Boolean {
        val path = navigator.path
        if (path != null) {
            val pos = path.finalPathPoint
            if (pos != null) {
                val distance = getDistanceSq(pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble())
                if (distance < 4.0) {
                    return true
                }
            }
        }

        return false
    }

    fun shouldAttack(entity: EntityLivingBase?): Boolean {
        return if (entity != null) {
            !world.isDaytime || entity.isInWater
        } else {
            false
        }
    }

    override fun attackEntityWithRangedAttack(target: EntityLivingBase, distanceFactor: Float) {
        val trident = EntityTrident(world, this, FItems.TRIDENT.stack)
        val x = target.posX - posX
        val y = target.entityBoundingBox.minY + (target.height / 3.0f) - trident.posY
        val z = target.posZ - posZ
        val v = sqrt(x * x + z * z)
        trident.shoot(x, y + v * 0.2f, z, 1.6f, 14 - (world.difficulty.id * 4).toFloat())
        playSound(Sounds.TRIDENT_THROW, 1.0f, 1.0f / rand.nextFloat() * 0.4f + 0.8f)
        world.spawnEntity(trident)
    }

    override fun setSwingingArms(swingingArms: Boolean) {
        dataManager[SWINGING_ARMS] = swingingArms
    }

    fun isSwingingArms() = dataManager[SWINGING_ARMS]

    fun getSwimAnimation(partialTicks: Float): Float {
        return lerp(partialTicks, lastSwimAnimation, swimAnimation)
    }

    companion object {
        private val DROWNING = EntityDataManager.createKey(EntityDrowned::class.java, DataSerializers.BOOLEAN)
        private val SWINGING_ARMS = EntityDataManager.createKey(EntityDrowned::class.java, DataSerializers.BOOLEAN)
        private val SWIMMING = EntityDataManager.createKey(EntityDrowned::class.java, DataSerializers.BOOLEAN)
    }
}